### PR TITLE
[containerapp] Hide containerapps and jobs env vars on create and update

### PR DESCRIPTION
**Pull request by p-bouchon**
---

This checklist is used to make sure that common guidelines for a pull request are followed.

### General Guidelines

- [x] Have you run `azdev style <YOUR_EXT>` locally? (`pip install azdev` required)
- [x] Have you run `python scripts/ci/test_index.py -q` locally? (`pip install wheel==0.30.0` required)

For new extensions:

- [ ] My extension description/summary conforms to the [Extension Summary Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/extensions/extension_summary_guidelines.md).

### About Extension Publish

There is a pipeline to automatically build, upload and publish extension wheels.  
Once your pull request is merged into main branch, a new pull request will be created to update `src/index.json` automatically.  
You only need to update the version information in file setup.py and historical information in file HISTORY.rst in your PR but do not modify `src/index.json`.

### CODE REVIEW

**Comment by yonzhan** [comment]
containerapp

File: src/containerapp/azext_containerapp/_transformers.py [hunk 1]
```diff
@@ -17,6 +17,38 @@ def transform_containerapp_output(app):
...
+
+def transform_sensitive_values_wrapper():
+
+    def transform_sensitive_values(response_json):
+        return clean_up_sensitive_values(response_json)
+
+    return transform_sensitive_values
+
+
+def transform_sensitive_values_list_output_wrapper():
```
Comment (Greedygre) [inline src/containerapp/azext_containerapp/_transformers.py:44]: `transform_sensitive_values_list_output_wrapper` is not used, should we remove it?

File: src/containerapp/azext_containerapp/_transformers.py [hunk 2]
```diff
@@ -17,6 +17,38 @@ def transform_containerapp_output(app):
...
+    if safe_get(response_json, "properties", "template") and "scale" in response_json["properties"]["template"]:
+        for rule in safe_get(response_json, "properties", "template", "scale", "rules", default=[]):
+            for (key, val) in rule.items():
+                if key != "name":
+                    val["metadata"] = dict((k, "") for k, v in val["metadata"].items())
+                    
+    return response_json
+
+
+def transform_sensitive_values_wrapper():
```
Comment (Greedygre) [inline src/containerapp/azext_containerapp/_transformers.py:36]: Should transform_sensitive_values_wrapper call `clean_up_sensitive_values`?

File: src/containerapp/azext_containerapp/_transformers.py [hunk 3]
```diff
@@ -3,7 +3,7 @@
 # Copyright (c) Microsoft Corporation. All rights reserved.
 # Licensed under the MIT License. See License.txt in the project root for license information.
 # --------------------------------------------------------------------------------------------
-
+from azure.cli.command_modules.containerapp._utils import safe_get
```
Comment (Greedygre) [inline src/containerapp/azext_containerapp/_transformers.py:6]: The **Azure Core CLI** (aka. [Azure CLI](https://learn.microsoft.com/en-us/cli/azure/what-is-azure-cli)) contains the key features for different Azure Resource provider. Starting from the 2.53.0 the Container App command has been added as a part of Azure the Core CLI
Currently the value of `azext.minCliCoreVersion` is not enough to import `safe_get`.
After [this PR](https://github.com/Azure/azure-cli-extensions/pull/6868) merged, update `"azext.minCliCoreVersion"` to `"2.53.0"`, you can import from `azure.cli.command_modules.containerapp`.
```suggestion
from ._utils import safe_get
```

**Comment by Greedygre** [comment]
Hi @p-bouchon 

Please fix the static analysis.
Running flake8 on extensions...
/mnt/vss/_work/1/s/src/containerapp/azext_containerapp/_transformers.py:8:1: E302 expected 2 blank lines, found 1
/mnt/vss/_work/1/s/src/containerapp/azext_containerapp/commands.py:16:55: W291 trailing whitespace

File: src/containerapp/azext_containerapp/commands.py [hunk 4]
```diff
@@ -13,18 +13,20 @@
...
+                            transform_sensitive_values_wrapper)
 
 
 def load_command_table(self, _):
     with self.command_group('containerapp') as g:
         g.custom_show_command('show', 'show_containerapp', table_transformer=transform_containerapp_output)
         g.custom_command('list', 'list_containerapp', table_transformer=transform_containerapp_list_output)
-        g.custom_command('create', 'create_containerapp', supports_no_wait=True, exception_handler=ex_handler_factory(), table_transformer=transform_containerapp_output)
-        g.custom_command('update', 'update_containerapp', supports_no_wait=True, exception_handler=ex_handler_factory(), table_transformer=transform_containerapp_output)
+        g.custom_command('create', 'create_containerapp', supports_no_wait=True, exception_handler=ex_handler_factory(), table_transformer=transform_containerapp_output, transform=transform_sensitive_values_wrapper())
```
Comment (Greedygre) [inline src/containerapp/azext_containerapp/commands.py:25]: Recommend `transform=transform_sensitive_values_wrapper`, no need `()`

File: src/containerapp/azext_containerapp/commands.py [hunk 5]
```diff
@@ -50,9 +52,9 @@ def load_command_table(self, _):
     with self.command_group('containerapp job') as g:
         g.custom_show_command('show', 'show_containerappsjob')
         g.custom_command('list', 'list_containerappsjob')
-        g.custom_command('create', 'create_containerappsjob', supports_no_wait=True, exception_handler=ex_handler_factory())
+        g.custom_command('create', 'create_containerappsjob', supports_no_wait=True, exception_handler=ex_handler_factory(), transform=transform_sensitive_values_wrapper())
```
Comment (Greedygre) [inline src/containerapp/azext_containerapp/commands.py:55]: Same as above.

File: src/containerapp/azext_containerapp/_transformers.py [hunk 6]
```diff
@@ -17,6 +17,26 @@ def transform_containerapp_output(app):
...
+        for rule in safe_get(response_json, "properties", "template", "scale", "rules", default=[]):
+            for (key, val) in rule.items():
+                if key != "name":
+                    val["metadata"] = dict((k, "") for k, v in val["metadata"].items())
+
+    return response_json
+
+
+def transform_sensitive_values_wrapper():
+    return clean_up_sensitive_values
```
Comment (Greedygre) [inline src/containerapp/azext_containerapp/_transformers.py:37]: I suggest merging the two methods `clean_up_sensitive_values `and `transform_sensitive_values_wrapper `together.
Seems no need to separate them.

**Comment by Greedygre** [comment]
Hi @Juliehzl 
Could you help to review this PR?

File: src/containerapp/azext_containerapp/_transformers.py [hunk 7]
```diff
@@ -17,6 +18,22 @@ def transform_containerapp_output(app):
     return result
 
 
+def clean_up_sensitive_values(response_json):
```
Comment (Greedygre) [inline src/containerapp/azext_containerapp/_transformers.py:21]: Please use a name start will `transform_`

**Comment by Greedygre** [comment]
Hi @p-bouchon 
Please add tests for properties.template.scale.rules[] metadata

File: src/containerapp/azext_containerapp/_transformers.py [hunk 8]
```diff
@@ -17,6 +18,22 @@ def transform_containerapp_output(app):
...
+        if "env" in container:
+            for env in container["env"]:
+                if env.get("value"):
+                    del env["value"]
+
+    if safe_get(response_json, "properties", "template") and "scale" in response_json["properties"]["template"]:
+        for rule in safe_get(response_json, "properties", "template", "scale", "rules", default=[]):
+            for (key, val) in rule.items():
+                if key != "name":
+                    val["metadata"] = dict((k, "") for k, v in val["metadata"].items())
```
Comment (Greedygre) [inline src/containerapp/azext_containerapp/_transformers.py:32]: Please add tests for hiding properties.template.scale.rules[] metadata

File: src/containerapp/azext_containerapp/commands.py [hunk 9]
```diff
@@ -13,18 +13,20 @@
...
-        g.custom_command('update', 'update_containerapp', supports_no_wait=True, exception_handler=ex_handler_factory(), table_transformer=transform_containerapp_output)
+        g.custom_command('create', 'create_containerapp', supports_no_wait=True, exception_handler=ex_handler_factory(), table_transformer=transform_containerapp_output, transform=transform_sensitive_values)
+        g.custom_command('update', 'update_containerapp', supports_no_wait=True, exception_handler=ex_handler_factory(), table_transformer=transform_containerapp_output, transform=transform_sensitive_values)
         g.custom_command('delete', 'delete_containerapp', supports_no_wait=True, confirmation=True, exception_handler=ex_handler_factory())
         g.custom_command('exec', 'containerapp_ssh', validator=validate_ssh)
-        g.custom_command('up', 'containerapp_up', supports_no_wait=False, exception_handler=ex_handler_factory())
+        g.custom_command('up', 'containerapp_up', supports_no_wait=False, exception_handler=ex_handler_factory(), transform=transform_sensitive_values)
```
Comment (Greedygre) [inline src/containerapp/azext_containerapp/commands.py:29]: please remove the `transform=transform_sensitive_values` for `up `is not needed.

Comment (Greedygre) [reply src/containerapp/azext_containerapp/commands.py:29]: I will remove it

Comment (Greedygre) [inline src/containerapp/azext_containerapp/commands.py:29]:
```suggestion
        g.custom_command('up', 'containerapp_up', supports_no_wait=False, exception_handler=ex_handler_factory())
```

**Review by Greedygre — APPROVED** [review]

File: src/containerapp/HISTORY.rst [hunk 10]
```diff
@@ -9,6 +9,10 @@ upcoming
 * 'az containerapp compose create': fixed an issue where the environment's resource group was not resolved from --environment when the input value was a resource id.
 * 'az containerapp replica count', returns the replica count of a container app
 * [Breaking Change] 'az containerapp job create': add default values for container app job properties --replica-completion-count, --replica-retry-limit, --replica-timeout, --parallelism, --min-executions, --max-executions, --polling-interval
+* 'az containerapp create': hide environment variables
+* 'az containerapp update': hide environment variables
+* 'az containerapp job create': hide environment variables
+* 'az containerapp job update': hide environment variables
```
Comment (Greedygre) [inline src/containerapp/HISTORY.rst:15]:
```suggestion
* 'az containerapp create/update': hide environment variables, scale rules metadata
* 'az containerapp job create/update': hide environment variables, scale rules metadata
```

File: src/containerapp/HISTORY.rst [hunk 11]
```diff
@@ -9,6 +9,8 @@ upcoming
 * 'az containerapp compose create': fixed an issue where the environment's resource group was not resolved from --environment when the input value was a resource id.
 * 'az containerapp replica count', returns the replica count of a container app
 * [Breaking Change] 'az containerapp job create': add default values for container app job properties --replica-completion-count, --replica-retry-limit, --replica-timeout, --parallelism, --min-executions, --max-executions, --polling-interval
+* 'az containerapp create/update': hide environment variables, scale rules metadata
+* 'az containerapp job create/update': hide environment variables, scale rules metadata
```
Comment (Juliehzl) [inline src/containerapp/HISTORY.rst:13]: add eventTriggerConfig for job

Comment (Greedygre) [inline src/containerapp/HISTORY.rst:13]:
```suggestion
* 'az containerapp job create/update': hide environment variables, scale rules metadata, eventTriggerConfig for job
```

File: src/containerapp/azext_containerapp/_transformers.py [hunk 12]
```diff
@@ -17,6 +18,26 @@ def transform_containerapp_output(app):
...
+
+    if safe_get(response_json, "properties", "configuration", "eventTriggerConfig") and "scale" in response_json["properties"]["configuration"]["eventTriggerConfig"]:
```
Comment (Juliehzl) [inline src/containerapp/azext_containerapp/_transformers.py:34]: add default

**Review by Juliehzl — APPROVED** [review]

**Comment by Greedygre** [comment]
Hi @yanzhudd 

Could you please help to review and merge this PR? Thanks!!

**Comment by yanzhudd** [comment]
please rerun test in live mode and upload the recording file.

**Review by yanzhudd — APPROVED** [review]

**Pull request merged**